### PR TITLE
NEW SMITE: Retcon, also refactors temporary_atom element

### PR DESCRIPTION
**Pull request by DATA-xPUNGED**
## About The Pull Request

Added a new smite called Retcon, which fades the victim into nothing after a configurable set of time (default being 5 seconds), they can still move and act and stuff while they're fading, which will probably lead to funny shenanigans
Also refactors the `temporary_atom` element into an atom-level proc called `fade_into_nothing()`
## Why It's Good For The Game

more admin abuse opportunities for the smite, and the element really wasn't doing anything element-ish, so why have it be an element in the first place?
## Changelog
:cl:
admin: Admins can now Retcon people out of existence.
refactor: refactored the temporary_atom element into the fade_into_nothing() proc
/:cl:

### CODE REVIEW

File: code/__HELPERS/fade_into_nothing.dm [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/// Deletes the atom with a little fading out animation after a specified time
+/atom/proc/fade_into_nothing(var/atom/target = src, life_time = 5 SECONDS, fade_time = 3 SECONDS)
+	addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(qdel), WEAKREF(target)), life_time, TIMER_DELETE_ME)
+	if (life_time > fade_time && fade_time > 0)
+		addtimer(CALLBACK(src, PROC_REF(fade_into_nothing_fade_out), WEAKREF(target), fade_time), life_time - fade_time, TIMER_DELETE_ME)
+
+/// Actually does the fade out, used by fade_into_nothing()
+/atom/proc/fade_into_nothing_animate(datum/weakref/target_ref, fade_time)
```
Comment (Jacquerel) [inline code/__HELPERS/fade_into_nothing.dm:8]:
```suggestion
		addtimer(CALLBACK(src, PROC_REF(fade_into_nothing_animate), WEAKREF(target), fade_time), life_time - fade_time, TIMER_DELETE_ME)

/// Actually does the fade out, used by fade_into_nothing()
/atom/proc/fade_into_nothing_animate(datum/weakref/target_ref, fade_time)
```

These two proc names don't match.

File: code/__HELPERS/fade_into_nothing.dm [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/// Deletes the atom with a little fading out animation after a specified time
+/atom/proc/fade_into_nothing(var/atom/target = src, life_time = 5 SECONDS, fade_time = 3 SECONDS)
```
Comment (Jacquerel) [inline code/__HELPERS/fade_into_nothing.dm:2]: If this is to be an atom proc, it probably doesn't need the first argument.
You could alternately make it a global proc and retain the first argument.

File: code/__HELPERS/fade_into_nothing.dm [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+	addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(qdel), WEAKREF(target)), life_time, TIMER_DELETE_ME)
+	if (life_time > fade_time && fade_time > 0)
+		addtimer(CALLBACK(src, PROC_REF(fade_into_nothing_fade_out), WEAKREF(target), fade_time), life_time - fade_time, TIMER_DELETE_ME)
+
+/// Actually does the fade out, used by fade_into_nothing()
+/atom/proc/fade_into_nothing_animate(datum/weakref/target_ref, fade_time)
+	var/atom/target = target_ref?.resolve()
```
Comment (Jacquerel) [inline code/__HELPERS/fade_into_nothing.dm:9]: We also don't need to pass the target as a weakref and resolve it if this is an atom proc, it should presumably just be `src`

File: code/modules/admin/smites/retcon.dm [hunk 4]
```diff
@@ -0,0 +1,12 @@
+/datum/smite/retcon
+	name = "Retcon"
+	var/timer //how long until the thing attacked by this smite gets deleted
+	var/fade_out_timer //the time it takes to actually do the fade out animation, the victim will always have some time still fully visible
```
Comment (Jacquerel) [inline code/modules/admin/smites/retcon.dm:4]:
```suggestion
	/// how long until the thing attacked by this smite gets deleted
	var/timer 
	/// the time it takes to actually do the fade out animation, the victim will always have some time still fully visible
	var/fade_out_timer 
```
Triple-slashed comments above var or proc declarations are picked up by our auto-documenter.
Use `//` for comments or stuff in the middle of a proc, `///` for documenting what a variable is for.

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by DrAmazing343** [comment]
Being able to Build Mode the temporary_atom effect is so peak actually

**Comment by distributivgesetz** [comment]
a `visible_message` about the target being removed from the story would be awesome for added RP tbh

**Comment by Jacquerel** [comment]
Your code still doesn't compile, let me know on Discord if you need some help untangling why as that seems more sensible than writing it for you using review notes 😅

**Comment by Ghommie** [comment]
I personally think retcon smite should also remove them from the crew manifest, remove any memory datum from other mobs' minds that reference them, delete their bank account, scrub them from antagonists objectives that reference the target and reopen up their job slot. We've plenty of smites and this is just some fade out, but I'm not going to pretend you'll even accomplish half of what I suggest.

**Comment by DATA-xPUNGED** [comment]
> Your code still doesn't compile, let me know on Discord if you need some help untangling why as that seems more sensible than writing it for you using review notes 😅

yeah i'm having issue with the weakref part, i'll ping you when i have free time

**Comment by DATA-xPUNGED** [comment]
> I personally think retcon smite should also remove them from the crew manifest, remove any memory datum from other mobs' minds that reference them, delete their bank account, scrub them from antagonists objectives that reference them and finally reopen up their job slot. We've plenty of smites and this is just some fade out, but I'm not going to pretend you'll even accomplish half of what I suggest.

I should make them scream with such timing that the scream gets cut off...

**Pull request closed without merging**